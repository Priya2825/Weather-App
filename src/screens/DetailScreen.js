import React, { useEffect, useState } from "react";
import {
  View,
  Text,
  Image,
  StyleSheet,
  Pressable,
  Dimensions,
} from "react-native";
import HorizontalFlatlist from "../../components/HorizontalFlatlist";
import axios from "axios";

const { width, height } = Dimensions.get("window");

function DetailScreen({ route, navigation }) {
  const item = route.params;
  const [hourlyForecast, setHourlyForecast] = useState([]);

  useEffect(() => {
    const fetchHourlyForecast = async () => {
      try {
        const response = await axios.get(
          "https://api.openweathermap.org/data/2.5/forecast",
          {
            params: {
              lat: item.coord.lat,
              lon: item.coord.lon,
              units: "metric",
              appid: "24f40166e842079cad21bf954642b936",
            },
          }
        );

        // Extract the first 5 time slots for simplicity
        const forecast = response.data.list.map((f) => ({
          time: f.dt_txt.split(" ")[1].slice(0, 5), // Extract only the time part
          temperature: Math.round(f.main.temp), // Round the temperature
          icon: `http://openweathermap.org/img/wn/${f.weather[0].icon}.png`, // Weather icon URL
          backgroundColor: "#FBFFF1", // Set a background color
          ...f,
        }));
        setHourlyForecast(forecast);
      } catch (error) {
        // console.error(error);
      }
    };

    fetchHourlyForecast();
  }, [item.coord.lat, item.coord.lon]);

  const HandleOnPress = () => {
    navigation.navigate("Report", { data: hourlyForecast, name: item.name });
  };

  return (
    <View style={[styles.container, { backgroundColor: item.backgroundColor }]}>
      <Text style={styles.location}>{item.name}</Text>
      <Image
        source={item.weatherIcon}
        style={styles.icon}
        testID="weather-icon"
      />
      <Text style={styles.temperature}>{item.main.temp}Â°C</Text>
      <Text style={styles.currentWeather}>{item.weather[0].main}</Text>

      <View style={styles.additionalInfoContainer}>
        <View style={styles.infoBlock}>
          <Text style={styles.infoLabel}>Wind</Text>
          <Text style={styles.infoValue}>{item.wind.speed}km/h</Text>
        </View>
        <View style={styles.infoBlock}>
          <Text style={styles.infoLabel}>Humidity</Text>
          <Text style={styles.infoValue}>{item.main.humidity}%</Text>
        </View>
      </View>
      <Pressable
        onPress={HandleOnPress}
        style={{
          padding: 20,
        }}
      >
        <Text>Show full report</Text>
      </Pressable>
      <HorizontalFlatlist hourlyForecast={hourlyForecast.slice(0, 5)} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#f5f5f5",
  },
  location: {
    fontSize: 32,
    fontWeight: "600",
    color: "#333",
    marginBottom: 15,
  },
  icon: {
    width: width * 0.5, // 50% of screen width
    height: height * 0.2, // 20% of screen height
    resizeMode: "contain",
    marginBottom: height * 0.01, // 1% of screen height
  },
  temperature: {
    fontSize: 64,
    fontWeight: "300",
    color: "#333",
    marginBottom: 10,
  },
  currentWeather: {
    fontSize: 20,
    color: "#666",
    marginBottom: 10,
  },
  additionalInfoContainer: {
    flexDirection: "row",
    justifyContent: "space-between",
    width: "80%",
    marginBottom: 30,
  },
  infoBlock: {
    alignItems: "center",
  },
  infoLabel: {
    fontSize: 18,
    color: "#666",
    marginBottom: 5,
  },
  infoValue: {
    fontSize: 22,
    fontWeight: "600",
    color: "#333",
  },
});

export default DetailScreen;
