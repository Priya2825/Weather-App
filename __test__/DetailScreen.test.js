import React from "react";
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from "@testing-library/react-native";
import renderer from "react-test-renderer";
import { NavigationContainer } from "@react-navigation/native";
import DetailScreen from "../src/screens/DetailScreen";
import axios from "axios";

// Mock the axios module
jest.mock("axios");
const mockedAxios = axios;

const mockNavigation = {
  navigate: jest.fn(),
};

const mockRoute = {
  params: {
    coord: { lat: 12.34, lon: 56.78 },
    name: "Test City",
    weatherIcon: "http://openweathermap.org/img/wn/01d.png",
    main: { temp: 25, humidity: 60 },
    weather: [{ main: "Clear", description: "clear sky", icon: "01d" }],
    wind: { speed: 10 },
    backgroundColor: "#FFFAF0",
  },
};

describe("DetailScreen", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(
        <NavigationContainer>
          <DetailScreen route={mockRoute} navigation={mockNavigation} />
        </NavigationContainer>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render the weather data and essential components", async () => {
    // Mock the API response
    mockedAxios.get.mockResolvedValue({
      data: {
        list: [
          {
            dt_txt: "2024-09-06 12:00:00",
            main: { temp: 25 },
            weather: [{ icon: "01d" }],
          },
        ],
      },
    });

    render(
      <NavigationContainer>
        <DetailScreen route={mockRoute} navigation={mockNavigation} />
      </NavigationContainer>
    );

    // Check if essential components are rendered
    expect(screen.getByText("Test City")).toBeTruthy();
    expect(screen.getByText("25°C")).toBeTruthy();
    expect(screen.getByText("Clear")).toBeTruthy();

    // Check if the weather icon is rendered
    const weatherIcon = screen.getByTestId("weather-icon");

    await waitFor(() => {
      expect(weatherIcon.props.source).toBe(
        "http://openweathermap.org/img/wn/01d.png"
      );
    });
  });

  it('should navigate to ReportScreen when "Show full report" button is pressed', async () => {
    // Mock the API response
    mockedAxios.get.mockResolvedValue({
      data: {
        list: [
          {
            dt_txt: "2024-09-06 12:00:00",
            main: { temp: 25 },
            weather: [{ icon: "01d" }],
          },
        ],
      },
    });

    render(
      <NavigationContainer>
        <DetailScreen route={mockRoute} navigation={mockNavigation} />
      </NavigationContainer>
    );

    // Press the button and check navigation
    fireEvent.press(screen.getByText("Show full report"));
    await waitFor(() => {
      expect(mockNavigation.navigate).toHaveBeenCalledWith("Report", {
        data: expect.any(Array),
        name: "Test City",
      });
    });
  });

  it("should fetch hourly forecast data and update state", async () => {
    // Mock the API response
    mockedAxios.get.mockResolvedValue({
      data: {
        list: [
          {
            dt_txt: "2024-09-06 12:00:00",
            main: { temp: 25 },
            weather: [{ icon: "01d" }],
          },
        ],
      },
    });

    render(
      <NavigationContainer>
        <DetailScreen route={mockRoute} navigation={mockNavigation} />
      </NavigationContainer>
    );

    await waitFor(() => {
      // Verify if the hourly forecast data is processed
      expect(screen.getByText("25°C")).toBeTruthy();
    });
  });
});
