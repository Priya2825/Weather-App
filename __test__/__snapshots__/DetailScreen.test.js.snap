// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailScreen renders correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f5f5f5",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      },
      {
        "backgroundColor": "#FFFAF0",
      },
    ]
  }
>
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 32,
        "fontWeight": "600",
        "marginBottom": 15,
      }
    }
  >
    Test City
  </Text>
  <Image
    source="http://openweathermap.org/img/wn/01d.png"
    style={
      {
        "height": 266.8,
        "marginBottom": 13.34,
        "resizeMode": "contain",
        "width": 375,
      }
    }
    testID="weather-icon"
  />
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 64,
        "fontWeight": "300",
        "marginBottom": 10,
      }
    }
  >
    25
    Â°C
  </Text>
  <Text
    style={
      {
        "color": "#666",
        "fontSize": 20,
        "marginBottom": 10,
      }
    }
  >
    Clear
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 30,
        "width": "80%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          {
            "color": "#666",
            "fontSize": 18,
            "marginBottom": 5,
          }
        }
      >
        Wind
      </Text>
      <Text
        style={
          {
            "color": "#333",
            "fontSize": 22,
            "fontWeight": "600",
          }
        }
      >
        10
        km/h
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <Text
        style={
          {
            "color": "#666",
            "fontSize": 18,
            "marginBottom": 5,
          }
        }
      >
        Humidity
      </Text>
      <Text
        style={
          {
            "color": "#333",
            "fontSize": 22,
            "fontWeight": "600",
          }
        }
      >
        60
        %
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "padding": 20,
      }
    }
  >
    <Text>
      Show full report
    </Text>
  </View>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsHorizontalScrollIndicator={false}
    stickyHeaderIndices={[]}
    style={
      {
        "width": "100%",
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
